Set `onion`, `vary` and `Link` headers in Next.js app

amoore108 opened the pull request:
Summary
======
- Adds server-side experiment headers to `vary` headers
- Adds `onion-location` header
- Adds `Link` header
- Adds `UNKNOWN` to `PageTypes` type, as we tend to just extend that type with `| 'Unknown'` anyway
- Sets `serverSideExperiments` at the `_app.page.tsx` level so all pages will get it as a prop without configuration
<img width="544" height="176" alt="Screenshot 2025-12-30 at 17 10 40" src="https://github.com/user-attachments/assets/2c5c5c59-34a0-42d9-bbdb-454499d6fe5b" />


Code changes
======
- Changes top-level "special" components `_app.page.tsx` and `_error.page.tsx` to a class component to be consistent with `_document.page.tsx`
- Moves header setting logic into their own files in the Next `utilities` folder to tidy up `_app`
- Detaches service-chain header from CSP header setting logic
- Moves `handleServerLogging` into `utilities` folder as well to tidy up `_document`


Developer Checklist
======

- [ ] **UX**
  - [ ] UX Criteria met (visual UX & screenreader UX)
- [ ] **Accessibility**
    - [ ] Accessibility Acceptance Criteria met
    - [ ] Accessibility swarm completed
    - [ ] Component Health updated
    - [ ] P1 accessibility bugs resolved 
    - [ ] P2/P3 accessibility bugs planned (if not resolved)
- [ ] **Security**
    - [ ] Security issues addressed
    - [ ] Threat Model updated
- [ ] **Documentation**
    - [ ] Docs updated (runbook, READMEs)
- [ ] **[Testing](#testing)** 
    - [ ] Feature tested on relevant environments
- [ ] **Comms** 
    - [ ] Relevant parties notified of changes


Testing
======

- [ ] Manual Testing required?
  - [ ] Local (`Ready-For-Test, Local`)
  - [ ] Test  (`Ready-For-Test, Test`)
  - [ ] Preview (`Ready-For-Test, Preview`)
  - [ ] Live (`Ready-For-Test, Live`)
- [ ] Manual Testing complete?
  - [ ] Local
  - [ ] Test
  - [ ] Preview
  - [ ] Live


## Additional Testing Steps
1. _List the steps required to test this PR._


Useful Links
======
 - [Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)
 - [Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)
- _Add Links to useful resources related to this PR if applicable._

amoore108 committed:
Move `sw.js` rewrite to `next.config`

amoore108 committed:
Update next.config.js

amoore108 committed:
Logic to set headers from `_document`

amoore108 committed:
Fix `tsc` error

amoore108 committed:
Remove `middleware`

amoore108 committed:
Check hostname

amoore108 committed:
Update _document.page.tsx

amoore108 committed:
Update index.ts

amoore108 committed:
Merge branch 'latest' into move-sw.js-rewrite-to-next.config

amoore108 committed:
Merge branch 'latest' into move-sw.js-rewrite-to-next.config

amoore108 committed:
Merge branch 'latest' into move-sw.js-rewrite-to-next.config

amoore108 committed:
Replace `cspHeaderResponse` with `DocumentContext` version

amoore108 committed:
Fetch toggles from `_document.page.tsx`

amoore108 committed:
Pass args as object

amoore108 committed:
Add `extractHeaders` function to global page props

amoore108 committed:
Pass `isNextJs: true` down

amoore108 committed:
Remove toggle check from `augmentWithDisclaimer`

The toggle is checked within the `Disclaimer` component itself, so this seems redundant to just not add the small block type to the page data

amoore108 committed:
Remove `toggles` from `getServerSideProps` in pages

amoore108 committed:
Remove `toggles` fetch from `getPageData`

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Remove `toggles` from old article fetcher to fix TS error

amoore108 committed:
fallback to `index` if no `id` on block

amoore108 committed:
Remove platform flags from `handleArticleRoute`

amoore108 committed:
Remove platform flag from `handleAvRoute`

amoore108 committed:
Remove platform flags from `downloads` page

amoore108 committed:
Remove platform flags from `live` page

amoore108 committed:
Remove platform flags from `send` page

amoore108 committed:
Remove platform flag from `watch` page

amoore108 committed:
Remove platform flags from catch-all route

amoore108 committed:
Remove platform flags from `wrapped` page

amoore108 committed:
Remove platform flags from `languages` page

amoore108 committed:
Remove unneeded import

amoore108 committed:
Remove `extractHeaders` from articles route

amoore108 committed:
Remove `extractHeaders` from av-embeds route

amoore108 committed:
Remove `extractHeaders` from downloads route

amoore108 committed:
Remove `extractHeaders` from live route

amoore108 committed:
Remove `extractHeaders` from send route

amoore108 committed:
Remove `extractHeaders` from watch route

amoore108 committed:
Remove `extractHeaders` from catch-all route

amoore108 committed:
Remove `extractHeaders` from wrapped route

amoore108 committed:
Remove duplicate `ctx.res?.setHeader`

amoore108 committed:
Move logic from _document and move to _app

amoore108 committed:
Fix type imports

amoore108 committed:
Remove `appProps` and add comment

amoore108 committed:
Use cached version of toggles

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Rename CSP header tests to be more reflective of intent

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Make toggles destruct more readable

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Add `service` to toggles fetch from `ctx`

amoore108 committed:
Pass `service` down to csp header function to simplify logic

amoore108 committed:
Move comment for clarity

amoore108 committed:
Get `service` from path instead of `query`

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Update augmentWithDisclaimer.test.ts

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Remove toggles fetch and common props from Offline page

amoore108 committed:
Merge branch 'latest' into nextjs-toggles-refactor

amoore108 committed:
Remove `async/await` from header setting

amoore108 committed:
Update _app.page.tsx

amoore108 committed:
Add `Unknown` page type

amoore108 committed:
Remove manual `Unknown` type unions

amoore108 committed:
Separate out header setting

amoore108 committed:
Move csp header function call logic into cspHeaderResponse function

amoore108 committed:
Fix tests

amoore108 committed:
Rename folder

amoore108 committed:
Move into separate utils folder

amoore108 committed:
Remove server experiments logic from articles route

amoore108 committed:
add tests

amoore108 committed:
Convert _app to class component to be consistent with _document

amoore108 committed:
Remove unneeded type cast

amoore108 committed:
Move `handleServerLogging` to separate file

amoore108 committed:
Update _app.page.tsx

amoore108 committed:
Update _app.page.tsx

amoore108 committed:
Update _app.page.tsx

amoore108 committed:
Update _app.page.tsx

amoore108 committed:
Add page type tests to `derivePageType`

amoore108 committed:
Use `page-type` header first if available

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Fix import order

amoore108 committed:
Update _app.page.tsx

amoore108 committed:
Update _app.page.tsx

amoore108 committed:
Update index.test.ts

amoore108 committed:
Convert _error page to class component

amoore108 committed:
Update _error.page.tsx

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Use `UNKNOWN_PAGE` const

amoore108 committed:
Mock url typo

amoore108 committed:
Check localhost domain using `some`

amoore108 committed:
Change instances of `'unknown'` page type to use const

amoore108 committed:
Update _error.page.tsx

amoore108 committed:
Make type import consistent

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
un-pluralise vary header function

amoore108 committed:
Add `Link` header function

amoore108 committed:
Fix typo in process.env setting in test

amoore108 committed:
Change `getAssetOrigins` to TS

amoore108 committed:
Move consts out of function

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

HarveyPeachey commented on ws-nextjs-app/utilities/derivePageType/index.test.ts:
```diff
@@ -1,39 +1,72 @@
 describe('derivePageType', () => {
-  it("should return UGC_PAGE if pathname includes 'send'", () => {
-    const pathname = '/burmese/send/xxxxxxxxx';
+  it('should strip our query params from the pathname', () => {
```
```suggestion
  it('should strip out query params from the pathname', () => {
```

amoore108 committed:
Update ws-nextjs-app/utilities/derivePageType/index.test.ts

Co-authored-by: Harvey Peachey <harvey.peachey@bbc.co.uk>

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

amoore108 committed:
Update responseHeaderTests.ts

amoore108 committed:
typo

amoore108 committed:
Extend `responseHeaderTests` integration tests

- CSP can't be fully tested as integration uses localhost for the domain and CSP headers aren't set on localhost

amoore108 committed:
Update responseHeaderTests.ts

amoore108 committed:
Merge branch 'latest' into set-vary-onion-headers-in-nextjs

eagerterrier reviewed (approved):
![T-Shirt-Image-2_1-1](https://user-images.githubusercontent.com/91352/76620293-c6fe3d00-6524-11ea-9cbd-8dcf3a156d29.jpg)
